/**
 * This code was GENERATED using the solita package.
 * Please DO NOT EDIT THIS FILE, instead rerun solita to update it or write a wrapper to add functionality.
 *
 * See: https://github.com/metaplex-foundation/solita
 */

type ErrorWithCode = Error & { code: number }
type MaybeErrorWithCode = ErrorWithCode | null | undefined

const createErrorFromCodeLookup: Map<number, () => ErrorWithCode> = new Map()
const createErrorFromNameLookup: Map<string, () => ErrorWithCode> = new Map()

/**
 * ResourceNotInitialized: 'The resource is not initialized'
 *
 * @category Errors
 * @category generated
 */
export class ResourceNotInitializedError extends Error {
  readonly code: number = 0x1770
  readonly name: string = 'ResourceNotInitialized'
  constructor() {
    super('The resource is not initialized')
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, ResourceNotInitializedError)
    }
  }
}

createErrorFromCodeLookup.set(0x1770, () => new ResourceNotInitializedError())
createErrorFromNameLookup.set(
  'ResourceNotInitialized',
  () => new ResourceNotInitializedError()
)

/**
 * InsufficientAmount: 'The amount provided for burning is greater than the amount of the held in the account'
 *
 * @category Errors
 * @category generated
 */
export class InsufficientAmountError extends Error {
  readonly code: number = 0x1771
  readonly name: string = 'InsufficientAmount'
  constructor() {
    super(
      'The amount provided for burning is greater than the amount of the held in the account'
    )
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, InsufficientAmountError)
    }
  }
}

createErrorFromCodeLookup.set(0x1771, () => new InsufficientAmountError())
createErrorFromNameLookup.set(
  'InsufficientAmount',
  () => new InsufficientAmountError()
)

/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 * @category generated
 */
export function errorFromCode(code: number): MaybeErrorWithCode {
  const createError = createErrorFromCodeLookup.get(code)
  return createError != null ? createError() : null
}

/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 * @category generated
 */
export function errorFromName(name: string): MaybeErrorWithCode {
  const createError = createErrorFromNameLookup.get(name)
  return createError != null ? createError() : null
}
